### ISSUE TITLE
[Bug] Qdrant with Aspire times out

### ISSUE THREAD

**Issue by JohnGalt1717** [microsoft/kernel-memory#913]
### Context / Scenario

If you spin up Qdrant using .net Aspire like this:

```
var vectorDb = builder
    .AddQdrant(ResourceConstants.VECTOR_DB,) //Passing ports also doesn't work
    .WithDataVolume();


```

Ensuring that your project has a WithReference, and then spin up Kernel Memory by creating a qdrantConfig from the connection string (which is verified to have valid http and apikey) and then passing .WithQdrantMemoryDb(qdrantConfig)

### What happened?

When you try and import a document you get a timeout within the qdrant memory db code.

A qdrant client works just fine.

It should just work.

It would be really nice if Kernel Memory wrapped a proper QdrantClient instead of using direct Http requests because then we could use the service provider to get a qdrant client and pass it into WithQdrantMemoryDb and know it works and everything would integrate nicely with Aspire.

Failing that, it would be nice if there were samples on how to integrate with Aspire with Kernel Memory for everything.

### Importance

I cannot use Kernel Memory

### Platform, Language, Versions

Windows, C#, .NET 9.0 RTM with Aspire 9.0.0

### Relevant log output

```shell
at System.Net.Http.HttpConnection.<SendAsync>d__56.MoveNext()
   at System.Net.Http.HttpConnectionPool.<SendWithVersionDetectionAndRetryAsync>d__50.MoveNext()
   at System.Net.Http.Metrics.MetricsHandler.<SendAsyncWithMetrics>d__5.MoveNext()
   at System.Net.Http.DiagnosticsHandler.<SendAsyncCore>d__10.MoveNext()
   at System.Net.Http.RedirectHandler.<SendAsync>d__4.MoveNext()
   at System.Net.Http.HttpClient.<<SendAsync>g__Core|83_0>d.MoveNext()
   at Microsoft.KernelMemory.MemoryDb.Qdrant.Client.QdrantClient`1.<ExecuteHttpRequestAsync>d__14.MoveNext()
   at Microsoft.KernelMemory.MemoryDb.Qdrant.Client.QdrantClient`1.<CreateCollectionAsync>d__3.MoveNext()
   at Microsoft.KernelMemory.Handlers.SaveRecordsHandler.<CreateIndexOnceAsync>d__19.MoveNext()
   at Microsoft.KernelMemory.Handlers.SaveRecordsHandler.<InvokeAsync>d__13.MoveNext()
   at Microsoft.KernelMemory.Pipeline.InProcessPipelineOrchestrator.<RunPipelineAsync>d__11.MoveNext()
   at Microsoft.KernelMemory.Pipeline.BaseOrchestrator.<ImportDocumentAsync>d__23.MoveNext()
   at Microsoft.KernelMemory.MemoryServerless.<ImportTextAsync>d__11.MoveNext()
   at Microsoft.KernelMemory.MemoryServerless.<ImportTextAsync>d__11.MoveNext()
```

**Comment by JohnGalt1717** [comment]
PS: I suspect this is because of how the http client is being created inside of the Qdrant client. It's breaking on the proxy stuff in aspire.

**Comment by dluc** [comment]
could you provide the full aspire builder code? it might be about the hostname to be set in the env var

**Comment by JohnGalt1717** [comment]
```
builder
    .AddProject<ArticleAnalysis>(ToolConstants.ARTICLE_ANALYSIS)
    .WithReference(vectorDb)
    .WithReference(ai)
    .WithReference(aiModel)
    .WithReference(aiEmbeddingModel)
    .WaitFor(aiModel)
    .WaitFor(aiEmbeddingModel);

```

I assume this is what you're referring to?

And the endpoint that is passed in is: http://localhost:16334 which is the grpc port set by aspire.

**Comment by JohnGalt1717** [comment]
I'm using it as a project, not a docker image. I'm using the nuget packages.

**Comment by JohnGalt1717** [comment]
Update: If i hard code the http URL it works. The problem is that because it isn't using the qdrant client and is httping everything, it can never work with what is passed by Aspire without hacking. (of which you have no way in production versus dev to know what the ports are)

**Comment by dluc** [comment]
Without access to code it's too hard to say, my guess is that some required configuration is not being passed to KM, so KM doesn't know which port to use when connecting to Qdrant, or simply might be using the default settings

**Comment by JohnGalt1717** [comment]
> Without access to code it's too hard to say, my guess is that some required configuration is not being passed to KM, so KM doesn't know which port to use when connecting to Qdrant, or simply might be using the default settings

See my comment that we posted at the same time.

**Comment by dluc** [comment]
> Update: If i hard code the http URL it works. The problem is that because it isn't using the qdrant client (GRPC) and is httping everything (slower), it can never work with what is passed by Aspire without hacking. (of which you have no way in production versus dev to know what the ports are)

in prod it's probably easier because aspire proxy is not preset. You could try disabling the proxy locally too, and just using the default ports, ie 6333 for qdrant

**Comment by JohnGalt1717** [comment]
> > Update: If i hard code the http URL it works. The problem is that because it isn't using the qdrant client (GRPC) and is httping everything (slower), it can never work with what is passed by Aspire without hacking. (of which you have no way in production versus dev to know what the ports are)
> 
> in prod it's probably easier because aspire proxy is not present. You could try disabling the proxy locally too, and just using the default ports, ie 6333 for qdrant

You can't do that without creating duplicate mappings because the default ones will also be created and then the connection string will still be the ones automatically created by aspire in the connection string information.  The connection string info from aspire correctly only has the GRPC endpoint, not the http endpoint since Qdrant prefers GRPC and that's the default for the qdrantClient which KM isn't using.

So basically the only way right now to use KM with Aspire is to just defeat Aspire entirely and pretend it doesn't exist and hard code everything.

**Comment by dluc** [comment]
Aspire like Kubernetes is an orchestrator, and shouldn't require do change the code being hosted specifically for it. Otherwise we'd had to change the code for 10 different orchestration engines :-)

I believe the issue here is simpler, e.g. what Aspire assumes about connection strings is not how every app works. For instance, KM doesn't use a "Connection String".

Anyway, I'll publish an Aspire example as soon as I find the time.

**Comment by JohnGalt1717** [comment]
Here's my KM setup:

```
builder.Services.AddKeyedSingleton(
    AIConstants.LOCAL_KEY,
    (sp, _) =>
    {
        var qdrantClient = sp.GetRequiredService<QdrantClient>();
        var qdrantConnectionString =
            builder.Configuration.GetConnectionString(ResourceConstants.VECTOR_DB)
            ?? throw new InvalidOperationException(
                "Vector Db connection string is invalid"
            );

        var qdrantConfig = VectorDbExtensions.ParseConnectionString(qdrantConnectionString);

        var memoryBuilder = new KernelMemoryBuilder()
            .WithQdrantMemoryDb("http://localhost:16333", qdrantConfig.APIKey)
            .WithOllamaTextEmbeddingGeneration(ollamaConfig, new GPT4oTokenizer())
            .With(
                new TextPartitioningOptions
                {
                    MaxTokensPerLine = 150,
                    MaxTokensPerParagraph = 500,
                    OverlappingTokens = 30,
                }
            )
            .WithOllamaTextGeneration(ollamaConfig, new GPT4oTokenizer())
            .With(
                new TextGenerationOptions
                {
                    Temperature = 0.0,
                    FrequencyPenalty = 1.4,
                    ResultsPerPrompt = 1,
                }
            );
        return memoryBuilder.Build();
    }
);
```

.WithQdrantMemoryDb("http://localhost:16333", qdrantConfig.APIKey) should be able to take a qdrantClient from sp.GetRequiredService<QdrantClient>() which was previously injected with builder.AddQdrantClient("aspireresourceid")

Then this would be aspire friendly and work.  Any other clients that KM uses that uses 3rd party clients should work the same way.

**Comment by JohnGalt1717** [comment]
> Aspire like Kubernetes is an orchestrator, and shouldn't require do change the code being hosted specifically for it. Otherwise we'd had to change the code for 10 different orchestration engines :-)
> 
> I believe the issue here is simpler, e.g. what Aspire assumes about connection strings is not how every app works. For instance, KM doesn't use a "Connection String".
> 
> Anyway, I'll publish an Aspire example as soon as I find the time.

The real issue is that even though I split the connection string into the endpoint and apikey and provide KM with that, the endpoint provided by aspire is the GRPC endpoint which KM doesn't use because KM is using a custom http qdrant client instead of the Qdrant provided client.

Here's the code that parses the connection string that the Aspire.Hosting.Qdrant library injects:

```
    internal static QdrantConfig ParseConnectionString(string? connectionString)
    {
        string endpoint;
        string? key = null;
        if (Uri.TryCreate(connectionString, UriKind.Absolute, out var uri))
            endpoint = uri.ToString();
        else
        {
            var connectionBuilder = new DbConnectionStringBuilder
            {
                ConnectionString = connectionString,
            };

            if (
                connectionBuilder.ContainsKey("endpoint")
                && Uri.TryCreate(
                    connectionBuilder["endpoint"].ToString(),
                    UriKind.Absolute,
                    out var serviceUri
                )
            )
            {
                endpoint = serviceUri.ToString();
            }
            else
            {
                throw new InvalidOperationException(
                    "endpoint is not part of the connection string."
                );
            }

            if (connectionBuilder.ContainsKey("key"))
                key = connectionBuilder["key"].ToString();
        }

        return new QdrantConfig { Endpoint = endpoint, APIKey = key ?? string.Empty };
    }

```

But I shouldn't have to do this, because KM should be using QdrantClient from their nuget package, not a custom rolled http client.

**Comment by dluc** [comment]
by the way, the qdrant client didn't exist when we started 3 years ago :-) We'll take a look, thanks

**Comment by JohnGalt1717** [comment]
Makes sense! Thanks!  I appreciate it!

**Comment by dluc** [comment]
Here's a working example:

```csharp
var builder = DistributedApplication.CreateBuilder(args);

var qdrant = builder.AddContainer("qdrant", "qdrant/qdrant")
    .WithHttpEndpoint(targetPort: 6333)
    .PublishAsContainer();

var qdrantEndpoint = qdrant.GetEndpoint("http");

builder.AddProject<Projects.Service>("kernel-memory")
    .WaitFor(qdrant)
    .WithEnvironment("KernelMemory__TextGeneratorType", "AzureOpenAIText")
    .WithEnvironment("KernelMemory__DataIngestion__EmbeddingGeneratorTypes__0", "AzureOpenAIEmbedding")
    .WithEnvironment("KernelMemory__DataIngestion__MemoryDbTypes__0", "Qdrant")
    .WithEnvironment("KernelMemory__Retrieval__EmbeddingGeneratorType", "AzureOpenAIEmbedding")
    .WithEnvironment("KernelMemory__Retrieval__MemoryDbType", "Qdrant")
    .WithEnvironment("KernelMemory__Services__Qdrant__Endpoint", qdrantEndpoint)
    .WithEnvironment("KernelMemory__Services__Qdrant__APIKey", "")
    .WithEnvironment("KernelMemory__Services__AzureOpenAIText__Endpoint", "https://....openai.azure.com/")
    .WithEnvironment("KernelMemory__Services__AzureOpenAIText__Deployment", "gpt-4o")
    .WithEnvironment("KernelMemory__Services__AzureOpenAIEmbedding__Endpoint", "https://....openai.azure.com/")
    .WithEnvironment("KernelMemory__Services__AzureOpenAIEmbedding__Deployment", "text-embedding-ada-002");

builder.Build().Run();
```
